### PR TITLE
Add iteration to TimeUnit

### PR DESCRIPTION
**Pull request by b-chu**
# What does this PR do?
**Stack**:
- #3019
- #3017
- #3013 ⬅
- #3012

⚠️ *Part of a stack created by [spr](https://github.com/ejoffe/spr). Do not merge manually using the UI - doing so may have unexpected results.*

# What issue(s) does this change relate to?

# Before submitting
- [ ] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [ ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [ ] Did you run the tests locally to make sure they pass?
- [ ] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

File: composer/core/time.py [hunk 1]
```diff
@@ -122,6 +124,20 @@ def __init__(
...
         self._value, self._unit = value, TimeUnit(unit)
 
+    @classmethod
+    def from_iteration(cls, iteration: int) -> Time:
+        """Create a :class:`Time` with units of :attr:`TimeUnit.ITERATION`.
+
+        Equivalent to ``Time(epoch, TimeUnit.EPOCH)``.
+
+        Args:
+            epoch (int): Number of epochs.
```
Comment (snarayan21) [inline composer/core/time.py:134]:
```suggestion
            iteration (int): Number of iterations.
```

File: composer/core/time.py [hunk 2]
```diff
@@ -122,6 +124,20 @@ def __init__(
                 raise TypeError(f'value {value} is of type {type(value)}. Units {unit} require integer values.')
         self._value, self._unit = value, TimeUnit(unit)
 
+    @classmethod
+    def from_iteration(cls, iteration: int) -> Time:
+        """Create a :class:`Time` with units of :attr:`TimeUnit.ITERATION`.
+
+        Equivalent to ``Time(epoch, TimeUnit.EPOCH)``.
```
Comment (snarayan21) [inline composer/core/time.py:131]: typo
-Brian

File: composer/core/time.py (hunk 1 above)
Comment (snarayan21) [reply composer/core/time.py:134]: another docstring typo below

File: composer/core/time.py [hunk 3]
```diff
@@ -391,37 +407,48 @@ def from_timestring(cls, timestring: str) -> Time:
...
         batch_in_epoch: Union[int, Time[int]] = 0,
         sample_in_epoch: Union[int, Time[int]] = 0,
         token_in_epoch: Union[int, Time[int]] = 0,
         total_wct: Optional[datetime.timedelta] = None,
+        iteration_wct: Optional[datetime.timedelta] = None,
         epoch_wct: Optional[datetime.timedelta] = None,
         batch_wct: Optional[datetime.timedelta] = None,
     ):
+        iteration = Time.from_input(iteration, TimeUnit.ITERATION)
+        if iteration.unit != TimeUnit.ITERATION:
```
Comment (snarayan21) [inline composer/core/time.py:448]: just for my understanding, when would this check not evaluate to True?

File: composer/core/time.py [hunk 4]
```diff
@@ -474,14 +511,17 @@ def __init__(
...
             'epoch': self.epoch.value,
             'batch': self.batch.value,
             'sample': self.sample.value,
             'token': self.token.value,
+            'epoch_in_iteration': self.epoch_in_iteration.value,
             'batch_in_epoch': self.batch_in_epoch.value,
             'sample_in_epoch': self.sample_in_epoch.value,
             'token_in_epoch': self.token_in_epoch.value,
             'total_wct': self.total_wct,
+            # 'iteration_wct': self.iteration_wct,
```
Comment (snarayan21) [inline composer/core/time.py:524]: should this be commented out?

File: composer/core/time.py [hunk 5]
```diff
@@ -493,14 +533,17 @@ def get_state(self) -> Dict[str, Union[Time[int], datetime.timedelta]]:
...
             'epoch': self.epoch,
             'batch': self.batch,
             'sample': self.sample,
             'token': self.token,
+            'epoch_in_iteration': self.epoch_in_iteration,
             'batch_in_epoch': self.batch_in_epoch,
             'sample_in_epoch': self.sample_in_epoch,
             'token_in_epoch': self.token_in_epoch,
             'total_wct': self.total_wct,
+            # 'iteration_wct': self.iteration_wct,
```
Comment (snarayan21) [inline composer/core/time.py:546]: here as well

**Review by snarayan21 — COMMENTED** [review]
some comments

File: composer/core/time.py (hunk 3 above)
Comment (b-chu) [reply composer/core/time.py:448]: You can pass in `iteration=Time(5, TimeUnit.EPOCH)`

File: composer/core/time.py (hunk 4 above)
Comment (b-chu) [reply composer/core/time.py]: Good catch!

**Review by snarayan21 — APPROVED** [review]
nice, lgtm

**Review by mvpatel2000 — CHANGES_REQUESTED** [review]
Holding per offline discussion

**Review by mvpatel2000 — APPROVED** [review]

**Comment by b-chu** [comment]
✓ Commit merged in pull request [#3017](https://github.com/mosaicml/composer/pull/3017)

**Pull request closed without merging**